### PR TITLE
Initial support for symbolic tracing with torch.fx allowing dynamic axes

### PR DESCRIPTION
**Pull request by michaelbenayoun**
# What does this PR do?

This PR enables to symbolic trace models without having to specify a fixed batch size and / or sequence length. 
To specify that an axis should not have a fixed shape, the value should be set to -1:

```python
traced = symbolic_trace(
    model,
    input_names=["input_ids", "attention_mask", "token_type_ids"],
    batch_size=-1, # means that batch size should be dynamic
    sequence_length=-1 # means that the sequence length should be dynamic
)
```

Currently, only the following models support dynamic axes:

- Albert
- Bert
- DistilBert
- MobileBert
- MegatronBert
- Electra

Traced models with dynamic axes cannot be retraced out of the box, to solve this 2 functions are provided:

- `prepare_for_retracing`: takes a traced model as input and outputs a model that can be retraced and some information that is needed to get back to dynamic axes
- `restore_after_retracing`: takes a retraced model and the information returned by `prepare_for_retracing` and set the dynamic axes back

Example:
```python
prepared, attributes = prepare_for_retracing(traced)
retraced = some_retracing_func(prepared)
final_model = restore_after_retracing(retraced, attributes)
``` 

To make things less tedious, `retrace_graph_with` is provided: it takes a traced model, and either a Tracer or a tracing function and performs the retracing:

```python
final_model = retrace_graph_with(traced, func=some_tracing_func)
```

Being able to retrace a model is important because that is how quantization can be done for instance.

Finally, this PR also provides sanity checks that validate that symbolic tracing is available for the model provided to `symbolic_trace`, and that it can be traced with dynamic axes if the user asks for it.

### CODE REVIEW

File: src/transformers/utils/fx.py [hunk 1]
```diff
@@ -295,7 +374,7 @@ def path_of_module(self, mod: nn.Module) -> str:
             if path is None:
                 path = self._insert_module_as_submodule(mod)
             if path is None:
-                raise NameError("module is not installed as a submodule")
+                raise NameError("Module is not installed as a submodule")
```
Comment (stas00) [inline src/transformers/utils/fx.py:377]: may be it'd help to print the name of the module here and another instance of the same below so that it gives context to the user? or do you think this is enough?

File: src/transformers/utils/fx.py [hunk 2]
```diff
@@ -318,11 +397,290 @@ def create_arg(self, a: Any) -> Argument:
...
+def _insert_encoder_sequence_length_node(gm: GraphModule) -> Tuple[GraphModule, Node]:
+    """
+    Helper function that inserts a node that retrieves the encoder sequence length dynamically from the input of the
+    model.
+    """
+    graph = gm.graph
+    input_names = gm.dummy_inputs.keys()
+    encoder_sequence_length_node = None
+    for node in graph.nodes:
+        if (node.op == "placeholder") and ("decoder" not in node.name) and (node.name in input_names):
```
Comment (stas00) [inline src/transformers/utils/fx.py:428]: probably dropping parentheses for consistency with the rest of the code? Did the autoformatter mess it up?

**Review by stas00 — COMMENTED** [review]
I don't pretend to understand the magic you have created, Michael, other than very superficially. Do we have anybody else at HF who works with fx and we could ask to review? Perhaps some of the recent hires?

Also you left a few TODOs - are they for later PRs?

Otherwise looking great - I left a few small suggestions if those resonate with you.

File: src/transformers/utils/fx.py [hunk 3]
```diff
@@ -318,11 +397,290 @@ def create_arg(self, a: Any) -> Argument:
...
+    graph.lint()
+    gm.recompile()
+
+    return gm
+
+
+def transform_to_dynamic_input(gm: GraphModule, is_retracing: bool = False) -> GraphModule:
+    """Transformation that enables traced models to perform inference on dynamic input shapes."""
+    graph = gm.graph
+    input_names = gm.dummy_inputs.keys()
```
Comment (mfuntowicz) [inline src/transformers/utils/fx.py:539]: Again, potentially wrap with `set(...)` as you're only checking for presence.

File: src/transformers/utils/fx.py [hunk 4]
```diff
@@ -318,11 +397,290 @@ def create_arg(self, a: Any) -> Argument:
...
+    static2dynamic = {}
+
+    # Inserting the nodes that will fetch the batch size and sequence lengths dynamically.
+    if gm.use_dynamic_batch_size:
+        _, batch_size_node = _insert_batch_size_node(gm)
```
Comment (mfuntowicz) [inline src/transformers/utils/fx.py:544]: Wouldn't make sense to update the reference to `gm` _(instead of throwing it away `_`)_ after it has been recompiled with the batch size node inserted?

File: src/transformers/utils/fx.py [hunk 5]
```diff
@@ -318,11 +397,290 @@ def create_arg(self, a: Any) -> Argument:
...
+        if gm.num_choices > 0:
+            with graph.inserting_after(batch_size_node):
+                static2dynamic[gm.static_batch_size * gm.num_choices] = graph.call_function(
+                    operator.mul, args=(batch_size_node, gm.num_choices)
+                )
+            graph.lint()
+            gm.recompile()
+
+    if gm.use_dynamic_sequence_length:
+        _, encoder_sequence_length_node = _insert_encoder_sequence_length_node(gm)
```
Comment (mfuntowicz) [inline src/transformers/utils/fx.py:555]: Same comment here about updating the `gm` reference after it has been recompiled.

File: src/transformers/utils/fx.py [hunk 6]
```diff
@@ -318,11 +397,290 @@ def create_arg(self, a: Any) -> Argument:
...
+            and torch.all(partial_position_ids == buffer)
+        ):
+            position_ids_buffer_name = name
+            gm.register_buffer(name, position_ids)
+
+    inserted = False
+    position_ids_node = None
+    for node in graph.nodes:
+        arg_names = [arg.name if isinstance(arg, Node) else "" for arg in node.args]
+        if position_ids_buffer_name in arg_names:
```
Comment (mfuntowicz) [inline src/transformers/utils/fx.py:514]: Why do we need to store an array with potential empty strings here? And then redo a full check over the array one more time for presence?

Would a generator do the job?
 `is_position_ids_present = any(args.name == position_ids_buffer_name if isinstance(args, Node) else False for arg in node.args):`

File: src/transformers/utils/fx.py (hunk 1 above)
Comment (michaelbenayoun) [reply src/transformers/utils/fx.py:377]: You're right, I think that if that happens the current solution was to put a pdb breakpoint, printing the name of the module can help to pinpoint the origin of the problem without having to do so.

Thanks for the suggestion!

File: src/transformers/utils/fx.py [hunk 7]
```diff
@@ -22,16 +24,93 @@
...
 from ..models.auto import get_values
+from ..models.bert import BertConfig
+from ..models.distilbert import DistilBertConfig
+from ..models.electra import ElectraConfig
+from ..models.gpt2 import GPT2Config
+from ..models.gpt_neo import GPTNeoConfig
+from ..models.gptj import GPTJConfig
+from ..models.megatron_bert import MegatronBertConfig
+from ..models.mobilebert import MobileBertConfig
+from ..models.t5 import T5Config
```
Comment (sgugger) [inline src/transformers/utils/fx.py:41]: We have worked really hard to decouple files of the library, so I am very much against importing all configs in this file. If you really need the config objects, the `CONFIG_MAPPING` will map a model type to its config, so you should use that instead.

File: src/transformers/utils/fx.py [hunk 8]
```diff
@@ -22,16 +24,93 @@
...
+)
+
+_SUPPORTED_MODELS_FOR_DYNAMIC_AXES = (
+    *_generate_supported_model_classes(AlbertConfig),
+    *_generate_supported_model_classes(BertConfig),
+    *_generate_supported_model_classes(DistilBertConfig),
+    *_generate_supported_model_classes(MobileBertConfig),
+    *_generate_supported_model_classes(ElectraConfig),
+    *_generate_supported_model_classes(MegatronBertConfig),
+)
```
Comment (sgugger) [inline src/transformers/utils/fx.py:116]: This would be better if the `_generate_supported_model_classes` took a model type ("albert", "bert" etc) instead of a live object. Also maybe put all supported model types in a list and all special models (like `GPT2DoubleHeadsModel`) in another list, then build those?

File: src/transformers/utils/fx.py [hunk 9]
```diff
@@ -318,11 +397,290 @@ def create_arg(self, a: Any) -> Argument:
...
+            old_position_ids_node = node.args[index]
+            old_position_ids_node.replace_all_uses_with(position_ids_node)
+
+    graph.lint()
+    gm.recompile()
+
+    return gm
+
+
+def transform_to_dynamic_input(gm: GraphModule, is_retracing: bool = False) -> GraphModule:
```
Comment (sgugger) [inline src/transformers/utils/fx.py]: Please use a more explicit variable name than `gm`.

File: src/transformers/utils/fx.py [hunk 10]
```diff
@@ -318,11 +397,290 @@ def create_arg(self, a: Any) -> Argument:
...
+    graph.lint()
+    gm.recompile()
+
+    gm.static2dynamic = static2dynamic
+    gm.dynamic2static = {v: k for (k, v) in static2dynamic.items()}
+
+    return gm
+
+
+def patch_arguments(gm: GraphModule, mapping: Union[Dict[Node, int], Dict[int, Node]]) -> GraphModule:
```
Comment (sgugger) [inline src/transformers/utils/fx.py]: Same here. `graph` is short enough for instance.

**Review by sgugger — COMMENTED** [review]
I am not super happy with the general design of the fx utils module: the goal is to decouple the modules at the maximum and avoid importing every single config in one file. Also I am unsure you really need the actual model classes anywhere, but just their names could be enough? I may be wrong but even then, maybe only dynamically getting the model class when needed via the `XXX_MAPPING` seems better to me than building variables with all models.

So I would suggest relying on class names as much as possible.

File: src/transformers/utils/fx.py (hunk 7 above)
Comment (michaelbenayoun) [reply src/transformers/utils/fx.py:41]: Actually if I understand correctly the MODEL_...._MAPPINGs are mapping a config class to a model class so that is the reason I was importing the config classes.
But thanks to your review I learned about CONFIG_MAPPING which maps model names to config classes. 
So doing `model name -> model config class -> model class` solves the issue with very small changes in the code, thanks!

File: src/transformers/utils/fx.py (hunk 9 above)
Comment (michaelbenayoun) [reply src/transformers/utils/fx.py]: I used gm because that is how GraphModule are commonly called in [torch.fx transformations](https://pytorch.org/docs/stable/fx.html), at least in the documentation.
I do not think that's an issue because:

- People looking at this will most likely be familiar with torch.fx
- There is the type annotation that makes it clear
- It is not possible to call it graph as Graph is actually another class (and a GraphModule contains a Graph attribute called graph)

**Comment by michaelbenayoun** [comment]
@sgugger What do you think of the way I've changed things regarding the config classes imports?

File: src/transformers/utils/fx.py [hunk 11]
```diff
@@ -22,16 +22,88 @@
...
+    *_generate_supported_model_classes("albert"),
+    *_generate_supported_model_classes("bert"),
+    *_generate_supported_model_classes("distilbert"),
+    *_generate_supported_model_classes("mobilebert"),
+    *_generate_supported_model_classes("electra"),
+    *_generate_supported_model_classes("megatron-bert"),
+    *_generate_supported_model_classes("gpt2"),
+    *_generate_supported_model_classes("gptj"),
+    *_generate_supported_model_classes("gpt_neo"),
+    *_generate_supported_model_classes("t5"),
```
Comment (sgugger) [inline src/transformers/utils/fx.py:92]: I would still make a list with those supported model types and then build `_REGULAR_SUPPORTED_MODELS` form it, just so we can make it easier to extend the support ;-)

File: src/transformers/utils/fx.py [hunk 12]
```diff
@@ -22,16 +22,88 @@
...
+_SUPPORTED_MODELS = _REGULAR_SUPPORTED_MODELS + _SPECIAL_SUPPORTED_MODELS
+
+_SUPPORTED_MODELS_FOR_DYNAMIC_AXES = (
+    *_generate_supported_model_classes("albert"),
+    *_generate_supported_model_classes("bert"),
+    *_generate_supported_model_classes("distilbert"),
+    *_generate_supported_model_classes("mobilebert"),
+    *_generate_supported_model_classes("electra"),
+    *_generate_supported_model_classes("megatron-bert"),
+)
```
Comment (sgugger) [inline src/transformers/utils/fx.py:104]: Same here

**Review by sgugger — APPROVED** [review]
Way better, thanks a lot!

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 13]
```diff
@@ -203,7 +203,7 @@ def unshape(x):
         q = q / math.sqrt(dim_per_head)  # (bs, n_heads, q_length, dim_per_head)
         scores = torch.matmul(q, k.transpose(2, 3))  # (bs, n_heads, q_length, k_length)
         mask = (mask == 0).view(mask_reshp).expand_as(scores)  # (bs, n_heads, q_length, k_length)
-        scores.masked_fill_(mask, -float("inf"))  # (bs, n_heads, q_length, k_length)
+        scores = scores.masked_fill(mask, -float("inf"))  # (bs, n_heads, q_length, k_length)
```
Comment (michaelbenayoun) [inline src/transformers/models/distilbert/modeling_distilbert.py:206]: @sgugger @LysandreJik it's a tiny modification but wanted to make sure it's okay on your end?

The reason for this is because tracing with torch.fx seems to ignore this op because it is inplace, making predictions wrong due to not masking the scores at all.

Comment (LysandreJik) [reply src/transformers/models/distilbert/modeling_distilbert.py:206]: I think this is fine, but it looks to be a bug in `torch.fx` ?

Comment (michaelbenayoun) [reply src/transformers/models/distilbert/modeling_distilbert.py:206]: Let me investigate on that and come back to you
